src/scene.rs

```diff
@@ -7,12 +7,108 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+use cgmath;
+use cgmath::prelude::*;
 use json;
-use std::slice;
+use std::{mem, slice};
 
 use {Camera, Gltf, Mesh, Skin};
 
-///  A node in the node hierarchy. When the node contains `skin`, all
+type Matrix3 = cgmath::Matrix3<f32>;
+type Matrix4 = cgmath::Matrix4<f32>;
+type Quaternion = cgmath::Quaternion<f32>;
+type Vector3 = cgmath::Vector3<f32>;
+
+/// 4x4 identity matrix.
+const IDENTITY: [f32; 16] = {
+    [1.0, 0.0, 0.0, 0.0, 
+     0.0, 1.0, 0.0, 0.0, 
+     0.0, 0.0, 1.0, 0.0, 
+     0.0, 0.0, 0.0, 1.0]
+};
+
+/// The transform for a `Node`.
+#[derive(Clone, Debug)]
+pub enum Transform {
+    /// 4x4 transformation matrix in column-major order.
+    Matrix {
+        /// 4x4 matrix.
+        matrix: [[f32; 4]; 4],
+    },
+
+    /// Decomposed TRS properties.
+    Decomposed {
+        /// `[x, y, z]` vector.
+        translation: [f32; 3],
+
+        /// `[x, y, z, w]` quaternion, where `w` is the scalar.
+        rotation: [f32; 4],
+
+        /// `[x, y, z]` vector.
+        scale: [f32; 3],
+    },
+}
+
+impl Transform {
+    /// Returns the matrix representation of this transform.
+    ///
+    /// If the transform is `Decomposed`, then the matrix is generated with the
+    /// equation `matrix = translation * rotation * scale`.
+    pub fn matrix(self) -> [[f32; 4]; 4] {
+        match self {
+            Transform::Matrix { matrix } => matrix,
+            Transform::Decomposed { translation, rotation, scale } => {
+                let t = Matrix4::from_translation(translation.into());
+                let r = Matrix4::from(Quaternion::from(rotation));
+                let s = Matrix4::from_nonuniform_scale(scale[0], scale[1], scale[2]);
+                (t * r * s).into()
+            },
+        }
+    }
+
+    /// Returns the decomposed representation of this transform.
+    ///
+    /// If the transform is `Matrix`, then the decomposition is extracted from the
+    /// matrix.
+    pub fn decomposed(self) -> ([f32; 3], [f32; 4], [f32; 3]) {
```

Comment:
This needs good tests.

Reply:
Time to write some tests.

Reply:
At the moment I'm testing by loading the sample models with `three`. All seems to be in order.

What kind of tests should be written? Sometimes the quaternion doesn't match the original but is an equivalent rotation.

Reply:
Hmm...for this kind of math I'd try to find a few examples with as many edge cases as possible (not exactly sure what this would be here), possible compute the expected result with an alternative implementation and use some "assert almost equals" function (for numeric issues).
Not sure about the quaternion - re-multiplying maybe?

Reply:
Or just dump the values from some tests with `three` into tests :)